metricbeat/module/kubernetes/util/metrics_cache.go

```diff
@@ -18,76 +18,84 @@
 package util
 
 import (
+	"context"
 	"sync"
 	"time"
 )
 
 // PerfMetrics stores known metrics from Kubernetes nodes and containers
-var PerfMetrics = NewPerfMetricsCache()
+var PerfMetrics = NewPerfMetricsCache(context.TODO())
 
 const defaultTimeout = 120 * time.Second
 
 var now = time.Now
-var sleep = time.Sleep
+var after = time.After
 
 // NewPerfMetricsCache initializes and returns a new PerfMetricsCache
-func NewPerfMetricsCache() *PerfMetricsCache {
+func NewPerfMetricsCache(ctx context.Context) *PerfMetricsCache {
 	return &PerfMetricsCache{
-		NodeMemAllocatable:   newValueMap(defaultTimeout),
-		NodeCoresAllocatable: newValueMap(defaultTimeout),
+		NodeMemAllocatable:   newValueMap(ctx, defaultTimeout),
+		NodeCoresAllocatable: newValueMap(ctx, defaultTimeout),
 
-		ContainerMemLimit:   newValueMap(defaultTimeout),
-		ContainerCoresLimit: newValueMap(defaultTimeout),
+		ContainerMemLimit:   newValueMap(ctx, defaultTimeout),
+		ContainerCoresLimit: newValueMap(ctx, defaultTimeout),
 	}
 }
 
 // PerfMetricsCache stores known metrics from Kubernetes nodes and containers
 type PerfMetricsCache struct {
-	mutex                sync.RWMutex
 	NodeMemAllocatable   *valueMap
 	NodeCoresAllocatable *valueMap
 
 	ContainerMemLimit   *valueMap
 	ContainerCoresLimit *valueMap
 }
 
-func newValueMap(timeout time.Duration) *valueMap {
-	return &valueMap{
-		values:  map[string]value{},
+func newValueMap(ctx context.Context, timeout time.Duration) *valueMap {
+	m := &valueMap{
+		values:  map[string]*value{},
 		timeout: timeout,
 	}
+	m.startWorkers(ctx)
+	return m
 }
 
 type valueMap struct {
-	sync.RWMutex
-	running bool
+	sync.Mutex
 	timeout time.Duration
-	values  map[string]value
+	values  map[string]*value
 }
 
 type value struct {
 	value   float64
 	expires int64
 }
 
+func (v *value) renew(timeout time.Duration) {
+	v.expires = now().Add(timeout).Unix()
+}
+
 // ContainerUID creates an unique ID for from namespace, pod name and container name
 func ContainerUID(namespace, pod, container string) string {
 	return namespace + "-" + pod + "-" + container
 }
 
 // Get value
 func (m *valueMap) Get(name string) float64 {
-	m.RLock()
-	defer m.RUnlock()
-	return m.values[name].value
+	return m.getWithDefault(name, 0)
 }
 
 // Get value
 func (m *valueMap) GetWithDefault(name string, def float64) float64 {
-	m.RLock()
-	defer m.RUnlock()
+	return m.getWithDefault(name, def)
+}
+
+func (m *valueMap) getWithDefault(name string, def float64) float64 {
+	m.Lock()
+	defer m.Unlock()
 	val, ok := m.values[name]
 	if ok {
+		val.renew(m.timeout)
```

Comment:
This is the main change in this PR, to renew the expiration time when the value is read.